bleep-core/src/scala/bleep/BleepExecutable.scala

```diff
@@ -0,0 +1,103 @@
+package bleep
+
+import bleep.internal.compat.OptionalCompatOps
+import bleep.internal.jvmRunCommand
+import bleep.logging.Logger
+import coursier.core.{ModuleName, Organization}
+
+import java.nio.file.{Files, Path}
+
+sealed trait BleepExecutable {
+  def command: Path
+  def args: List[String]
+  def whole: List[String] = command.toString +: args
+}
+
+object BleepExecutable {
+  private val BleepMain = "bleep.Main"
+  private val ScalaTestRunner = "org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner"
+  private val SbtForkMain = "sbt.ForkMain"
+
+  sealed trait Binary extends BleepExecutable {
+    override def args: List[String] = Nil
+  }
+
+  case class CurrentBinary(command: Path) extends Binary
+  case class DownloadedBinary(command: Path) extends Binary
+  case class CurrentJava(command: Path, args: List[String]) extends BleepExecutable
+  case class DownloadedJava(command: Path, args: List[String]) extends BleepExecutable
+
+  def getCommand(resolver: CoursierResolver, pre: Prebootstrapped, forceJvm: Boolean): BleepExecutable =
+    findCurrentBleep(pre.logger).getOrElse {
+      val latestRelease = model.BleepVersion.current.latestRelease
+      OsArch.current match {
+        case image: OsArch.HasNativeImage if !forceJvm =>
+          pre.logger.warn(s"couldn't determine name of current Bleep executable. Setting up version ${latestRelease.value}")
+          val bin = FetchBleepRelease(latestRelease, pre.cacheLogger, pre.ec, image).orThrow
+          DownloadedBinary(bin)
+        case other =>
+          if (forceJvm) pre.logger.info(s"Setting up Bleep through a JVM as requested")
+          else pre.logger.warn(s"There is no published graalvm native-image for $other. Setting up Bleep through a JVM")
+
+          val bleepCliDep = model.Dep.ScalaDependency(Organization("build.bleep"), ModuleName("bleep-cli"), latestRelease.value, fullCrossVersion = false)
+          val resolvedBleep = resolver.force(
+            Set(bleepCliDep),
+            model.VersionCombo.Jvm(model.VersionScala.Scala213),
+            s"resolving bleep ${latestRelease.value} from maven central"
+          )
+
+          DownloadedJava(
+            pre.resolvedJvm.forceGet.javaBin,
+            jvmRunCommand.cmdArgs(jvmOptions = Nil, cp = resolvedBleep.jars, main = BleepMain, args = Nil)
+          )
+      }
+    }
+
+  implicit class IndexOfOpt(haystack: Array[String]) {
+    def indexOfOpt(needle: String): Option[Int] =
+      haystack.indexOf(needle) match {
+        case -1 => None
+        case n  => Some(n)
+      }
+  }
+
+  def findCurrentBleep(logger: Logger): Option[BleepExecutable] = {
+    def complain(msg: String): None.type = {
+      logger.warn(s"Problem while discovering how to run the bleep you started: $msg. Falling back to downloading latest release.")
+      None
+    }
+    val currentInfo = ProcessHandle.current.info()
```

Comment:
🎉

Reply:
This was a really great find! No idea why all those Google searches i did in the past didn't bring it up.

I have some time for bleep today, i think I'll merge this in one form or another.

The experience has some rough edges, and i feel like the whole thing is exactly on the edge of the complexity budget of what should go in, but we need this somehow. 

The integration tests in particular are difficult to understand and brittle.

Also things like how bloop behaves when the sourcegen is incorrectly setup is not optimal. Invalidation is also not on point, bloop does not always pick up changes to sourcegen scripts.

Reply:
Does this imply that it'd be preferable longer term for bleep to own the code-generation logic ?

Reply:
Maybe! I think it's best that the logic for running these things live in bloop, because it can know better when to run them. I think I've been observing bugs in the implementation there, and those can be reported and fixed.

Reply:
👍